### ISSUE TITLE
Null-conditional assignment

### ISSUE THREAD

**Issue by RikkiGibson** [dotnet/csharplang#6045]
# Null-conditional assignment

* [x] Proposed
* [ ] Prototype: [Not Started](https://github.com/PROTOTYPE_OWNER/roslyn/BRANCH_NAME)
* [ ] Implementation: [Not Started](https://github.com/dotnet/roslyn/BRANCH_NAME)
* [ ] Specification: Included

## Summary
[summary]: #summary

Permits assignment to occur conditionally within a `a?.b` or `a?[b]` expression.

Championed issue: https://github.com/dotnet/csharplang/issues/2883

```cs
using System;

class C
{
    public object obj;
}

void M(C? c)
{
    c?.obj = new object();
}
```

```cs
using System;

class C
{
    public event Action E;
}

void M(C? c)
{
    c?.E += () => { Console.WriteLine("handled event E"); };
}
```

```cs
void M(object[]? arr)
{
    arr?[42] = new object();
}

```
## Motivation
[motivation]: #motivation

A variety of motivating use cases can be found in the championed issue. Major motivations include:
1. Parity between properties and `Set()` methods.
2. Attaching event handlers in UI code.

## Detailed design
[design]: #detailed-design

- The right side of the assignment is only evaluated when the receiver of the conditional access is non-null.
```cs
// M() is only executed if 'a' is non-null.
// note: the value of 'a.b' doesn't affect whether things are evaluated here.
a?.b = M();
```

- All forms of compound assignment are allowed.
```cs
a?.b -= M(); // ok
a?.b += M(); // ok
// etc.
```

- If the result of the expression is used, the expression's type must be known to be of a value type or a reference type. This is consistent with existing behaviors on conditional accesses.
```cs
class C<T>
{
    public T? field;
}

void M1<T>(C<T>? c, T t)
{
    (c?.field = t).ToString(); // error: 'T' cannot be made nullable.
    c?.field = t; // ok
}
```

- Conditional access expressions are still not lvalues, and it's still not allowed to e.g. take a `ref` to them.
```cs
M(ref a?.b); // error
```

- It is not allowed to ref-assign to a conditional access. The main reason for this is that the only way you would conditionally access a ref variable is a ref field, and ref structs are forbidden from being used in nullable value types. If a valid scenario for a conditional ref-assignment came up in the future, we could add support at that time.
```cs
ref struct RS
{
    public ref int b;
}

void M(RS a, ref int x)
{
  a?.b = ref x; // error: Operator '?' can't be applied to operand of type 'C'.
}
```

- It's not possible to e.g. assign to conditional accesses through deconstruction assignment. We anticipate it will be rare for people to want to do this, and not a significant drawback to need to do it over multiple separate assignment expressions instead.
```cs
(a?.b, c?.d) = (x, y); // error
```

### Specification
The *null conditional assignment* grammar is defined as follows:

```antlr
null_conditional_assignment
    : null_conditional_member_access assignment_operator expression
    : null_conditional_element_access assignment_operator expression
```
See [§11.7.7](https://github.com/dotnet/csharpstandard/blob/draft-v6/standard/expressions.md#1177-null-conditional-member-access) and [§11.7.11](https://github.com/dotnet/csharpstandard/blob/draft-v6/standard/expressions.md#11711-null-conditional-element-access) for reference.

When the *null conditional assignment* appears in an expression-statement, its semantics are as follows:
- `P?.A = B` is equivalent to `if (P is not null) P.A = B;`, except that `P` is only evaluated once.
- `P?[A] = B` is equivalent to `if (P is not null) P[A] = B`, except that `P` is only evaluated once.

Otherwise, its semantics are as follows:
- `P?.A = B` with result type `T` is equivalent to `(P is null) ? (T?)null : (P.A = B)`, except that `P` is only evaluated once.
- `P?[A] = B` with result type `T` is equivalent to `(P is null) ? (T?)null : (P[A] = B)`, except that `P` is only evaluated once.

### Implementation
The grammar in the standard currently doesn't correspond strongly to the syntax design used in the implementation. We expect that to remain the case after this feature is implemented. The syntax design in the [implementation](https://github.com/dotnet/roslyn/blob/09408ab8a29e03caddfb11f29328c05169ac7cde/src/Compilers/CSharp/Portable/Syntax/Syntax.xml#L583-L607) isn't expected to actually change--only the way it is used will change. For example:

```mermaid
graph TD;
subgraph ConditionalAccessExpression
  whole[a?.b = c]
end
subgraph  
  direction LR;
  subgraph Expression
    whole-->a;
  end
  subgraph OperatorToken
    whole-->?;
  end
  subgraph WhenNotNull
    whole-->whenNotNull[.b = c];
    whenNotNull-->.b;
    whenNotNull-->=;
    whenNotNull-->c;
  end
end
```

### Complex examples
```cs
class C
{
    ref int M() => /*...*/;
}

void M(C? c)
{
    c?.M() = 42; // equivalent to:
    if (c is not null)
        c.M() = 42;
}
```

```cs
M(a?.b?.c = d); // equivalent to:
M(a is null
    ? null
    : (a.b is null
        ? null
        : (a.b.c = d));
```

```cs
a?.b = c?.d = e?.f; // equivalent to:
a?.b = (c?.d = e?.f);
```

```cs
a?.b ??= c; // equivalent to:
if (a is not null)
{
    if (a.b is null)
    {
        a.b = c;
    }
}
```

## Drawbacks
[drawbacks]: #drawbacks

The choice to keep the assignment within the conditional access introduces some additional work for the IDE, which has many code paths which need to work backwards from an assignment to identifying the thing being assigned.

## Alternatives
[alternatives]: #alternatives

We could instead make the `?.` syntactically a child of the `=`. This makes it so any handling of `=` expressions needs to become aware of the conditionality of the right side in the presence of `?.` on the left. It also makes it so the structure of the syntax doesn't correspond as strongly to the semantics.

## Unresolved questions
[unresolved]: #unresolved-questions

**Comment by CyrusNajmabadi** [comment]
> We could instead make the ?. syntactically a child of the =. This makes it so any handling of = expressions needs to become aware of the conditionality of the right side in the presence of ?. on the left. It also makes it so the structure of the syntax doesn't correspond as strongly to the semantics.

i would not do this :)  Let's be consistent with the way we already do conditionals today.   

On the IDE, we'll just need to update some of our conditional helpers we have today, and we'll have to audit all code that handles assignments.  Seems doable.

**Comment by quinmars** [comment]
I'd expect

```csharp
M(a?.b?.c = d);
```

be equivalent to:

```csharp
M(a is null
    ? d
    : (a.b is null
        ? d
        : (a.b.c = d));
```

so that for any expression, be it `a.b` or `a?.b`:

```
var z = (any_expression = d);
```

`z` would always be of the type of `d` and had the value of `d`. I know that would go against the lazy evaluation. And probably that is more important in real code, inparticular since the value of an assignment is rarely used.  But for a carefree reader, it could cause surprise.

**Comment by CyrusNajmabadi** [comment]
I would expect that if 'a' was null that no further work was done.  That's the prime use case as otherwise we'll execute expensive work that will be thrown away.  Having this work differently fro an argument vs an expression-statement would just be enormously weird at that point.  If someone ends up writing `M(a?.b?.c = d);` my expectation woudl be that they're thinking htey can get null as `a` being null would mean no work should be done at all.

I would also be entirely ok with the expression form of this being disallowed, and only allowing the statement form.

**Comment by RikkiGibson** [comment]
> ```
> var z = (any_expression = d);
> ```
> 
> `z` would always be of the type of `d` and had the value of `d`.

`d` is converted to the type of `any_expression` here, and the result type of the `=` expression is the same as the type of `any_expression`.

If we were to adopt the semantics you described @quinmars, then in `a?.b?.c = d`, we would probably want to convert `d` to the type of `a.b.c` and use that as the result type of the expression.

`a?.SetX(M())` has semantics where `M()` is only evaluated if a is not null. I think if anyone ended up moving from that form to `a?.X = M()` due to improved ergonomics here, they'd be disappointed if it ended up having different semantics.

**Comment by quinmars** [comment]
Nevermind, you are both correct. The lazy evaluation is a very important part of the feature and shouldn't be sacrified for my first impression when seeing this code. I find the symmetry argument (`a?.SetX(M())` vs. `a?.X = M()`) most compelling. Fortunately, the type system will catch wrong assumptions (`z` has the same type as `d`) easily - at least for value types. Anyway, the expression form of the assignment is seldom used.

**Comment by jnm2** [comment]
> I would also be entirely ok with the expression form of this being disallowed, and only allowing the statement form.

This has consequences for lambda return type inference, so it would leave a weird back-compat wart if this restriction was shipped and then loosened later.

**Comment by CyrusNajmabadi** [comment]
Let me rephrase: i would prefer to disallow one of the forms *over* having the two forms have different semantics :)  

If hte two forms have the same semantics, then there's no problem and i wouldn't disallow either of them.

**Comment by jnm2** [comment]
I agree, having the expression mean something different when directly a statement would be frightening.

I also initially expected `a = b?.c = d` to mean `temp = d; b?.c = temp; a = temp;`, but that's incompatible with my other strong expectation that `d` is not evaluated when `b` is null. So the first expectation has been revised.

**Comment by michael-hawker** [comment]
Awesome to see this moving forward! 🎉 Just wanted to bubble up our use case from the Windows Community Toolkit where we have to do event hooks all the time (_[from this original comment here](https://github.com/dotnet/csharplang/issues/2883#issuecomment-614294269)_):

>We [do this all the time in the toolkit for templated controls](https://github.com/windows-toolkit/WindowsCommunityToolkit/blob/068faf491fcb040ef7fafc15ce18710946eadd57/Microsoft.Toolkit.Uwp.UI.Controls/TabView/TabView.cs#L92-L115), where we need to register event handlers:


```cs
            if (_tabItemsPresenter != null)
            {
                _tabItemsPresenter.SizeChanged -= TabView_SizeChanged;
            }

            if (_tabScroller != null)
            {
                _tabScroller.Loaded -= ScrollViewer_Loaded;
            }

            _tabContentPresenter = GetTemplateChild(TabContentPresenterName) as ContentPresenter;
            _tabViewContainer = GetTemplateChild(TabViewContainerName) as Grid;
            _tabItemsPresenter = GetTemplateChild(TabsItemsPresenterName) as ItemsPresenter;
            _tabScroller = GetTemplateChild(TabsScrollViewerName) as ScrollViewer;

            if (_tabItemsPresenter != null)
            {
                _tabItemsPresenter.SizeChanged += TabView_SizeChanged;
            }

            if (_tabScroller != null)
            {
                _tabScroller.Loaded += ScrollViewer_Loaded;
            }
```

> It'd be nice to condense this down to this:


```cs
            _tabItemsPresenter?.SizeChanged -= TabView_SizeChanged;
            _tabScroller?.Loaded -= ScrollViewer_Loaded;

            _tabContentPresenter = GetTemplateChild(TabContentPresenterName) as ContentPresenter;
            _tabViewContainer = GetTemplateChild(TabViewContainerName) as Grid;
            _tabItemsPresenter = GetTemplateChild(TabsItemsPresenterName) as ItemsPresenter;
            _tabScroller = GetTemplateChild(TabsScrollViewerName) as ScrollViewer;

            _tabItemsPresenter?.SizeChanged += TabView_SizeChanged;
            _tabScroller?.Loaded += ScrollViewer_Loaded;
```

**Comment by Duranom** [comment]
Question concerning the `P?.A = B;` to be equivalent to `if (P is not null) { P.A = B; }`.
Is this the case for class types and structs? Or only for structs?
Because for classes I would expect `P?.A = B;` to be equivalent to `if (P is {} __p) { __p.A = B; }` 

For structs it also feels a bit weird to see `P?.A = B;` being equivalent to `if (P is not null) { P.A = B; }`.

Reason I ask is that it is easy enough to get race conditions with this, especially with UI or those that handle parallelism and it can easily assign incorrectly in cases when not aware.

**Comment by CyrusNajmabadi** [comment]
As the OP says:

> P?.A = B is equivalent to if (P is not null) P.A = B;, except that P is only evaluated once.

**Comment by Duranom** [comment]
My bad, the repeated parts later in example on put on wrong footing, as it feels far from equivalent.
But still wondering how it will go with structs, are those magically going to be referenced? Or will there be a need for an analyzer to warn if the struct is not done by reference?

**Comment by CyrusNajmabadi** [comment]
> But still wondering how it will go with structs, are those magically going to be referenced? Or will there be a need for an analyzer to warn if the struct is not done by reference?

**Comment by RikkiGibson** [comment]
I think we probably have to block assignments to nullable structs for the same reason we block the following: [SharpLab](https://sharplab.io/#v2:EYLgtghglgdgPgAQMwAIDOAXATgVwMYYoDKAsAFADe5KNKydAjAAwoAKWA9gA4oUoDmAUwwBudMLEBfctLLl6CAEwoAwr2q0FAFhQBZABREA/OgCU6srSvoAdADUIAGxyCb7bigC8KAEQQfYgD0gSiCWJxYqkQMAGwMiiCqEDAwHIRgHAAmUABmAJ4oGAAWgnQA7CgAbk4uKBw5KADkxvY1go0owIJ4EDhopVCEUGgoqYQQVRBYUBDAjoIaNLKSQA===)

```cs
public struct S
{
    public string Prop { get; set; }
}

public class C {
    public void M(S? s) {
        s.Value.Prop = "a"; // error CS1612: Cannot modify the return value of 'S?.Value' because it is not a variable
    }
}
```

**Comment by Casey-Hofland** [comment]
> `s?.Prop = "a"` would be the same error.

Just as well that's easy enough to include. A simple check to see if you're dealing with a Nullable<> and the condition evaluates to:

```
s?.Prop = "a"; // equivalent to:
if (s is not Nullable) // pseudo code
{
    // Do nullable reference type check
}
else if (s.HasValue)
{
    s = s.Value with { Prop = "a" }; // see [with expression](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/with-expression)
}
```

**Comment by RikkiGibson** [comment]
I think it is not legal, because we don't define `??` as a conditional operation (i.e. something which is grouped under the right-side of a `?` operator, such as `.M()` in `a?.M()` or `[i]` in `a?[i]`). So you end up back in the case the language says you are trying to assign to something which is not an lvalue.

**Comment by RikkiGibson** [comment]
Regarding https://github.com/dotnet/csharplang/issues/6045#issuecomment-1242414606

Thanks for the very interesting discussion. It feels like nullable value types and reference types are maybe experiencing an analogous issue here. [SharpLab](https://sharplab.io/#v2:EYLgtghglgdgNAFxBAzggPgWAFAGED8ABALICmCAFgPYAmAjABQGGwAOArggJQ4DeOhQYQACAdhYwOCQgHcolQr0IAzKKQA2NQgF5CAFgDMhAL4BuQgHoLsiACcYsAOYhCAcjadXhSAE9CwUkIYdnV1HGMcHABlIjJKWgAmBhiJKR5sfmwhEXEPBAA6ADUIdXZAuQUlVQ0tXUMTcysbeycXADkQ9QhgdUCANxKywgQfVkDff0Dg0PzwyOxbUgBjKlstXEURI1hpas1zCOw0W3Yl6SjN4W2YXbV9kxwgA)

```cs
C? Method1(C? input)
{
    return input with { field = 43 }; // warning: 'input' may be null
}

S? Method2(S? input)
{
    return input.Value with { field = 43 }; // warning: Nullable value type may be null.
}

record C { public int field; }
struct S { public int field; }
```

In the original proposal, we are improving the ergonomics of "update this thing if it not null", but we are maybe leaving behind "create an *updated version* of this thing if it is not null". The closest thing to what we really want is still `input == null ? null : input with { ... }`.

I'm not certain whether/how to address but I think we should keep an eye on it and see if people keep hitting it.

**Comment by jnm2** [comment]
A discussion of `with` on nullable things has also started here: https://github.com/dotnet/csharplang/discussions/6575

**Comment by CyrusNajmabadi** [comment]
afaict, this is not null-conditional-assignment.  It doesn't match teh grammar for that feature.  This is normal assignment.  In this case, the LHS doesn't produce LValues to assign into, so it's simply illegal and doesn't compile.

**Comment by SF-Simon** [comment]
C# is still too conservative.

If you want to read, you can write an "if" according to your own ideas.

I have noticed that the information related to this writing method has not been accepted for more than five years, which is really a magic thing.

I don't understand that most other development languages accept the writing method. Why does C # reject it? How much consideration is needed to provide a convenient writing method?

As Swift `if (var xx=nullable) {}` is written in this way. Why do we have to use a descriptive grammar to increase the difficulty? `if(nullable is not null and var xxx)`... It is very cumbersome and backward in thinking.

**Comment by HaloFour** [comment]
@SF-Simon 

If you have specific syntax proposals that you would like to suggest based on Swift feel free to open new discussions.

Swift does treat optionals differently than C#, though, in the sense that you are more or less forced to unwrap them into a different identifier.  You don't have to do that in C#, and the idiomatic null check `x != null` or `x is not null` is generally more concise.

**Comment by SF-Simon** [comment]
> @SF-Simon
> 
> If you have specific syntax proposals that you would like to suggest based on Swift feel free to open new discussions.
> 
> Swift does treat optionals differently than C#, though, in the sense that you are more or less forced to unwrap them into a different identifier. You don't have to do that in C#, and the idiomatic null check `x != null` or `x is not null` is generally more concise.

Hi, @HaloFour , sorry, maybe I didn't express myself clearly.

I don't mean to belittle C#, this is just an example. (I am using multiple development languages, all of which I like.)

As for the syntax of Swfit development language, I think it is very excellent because it simplifies many processes and is very easy to understand.

For example, Swift's weak reference keyword "_weak_" is much simpler and easier to understand than C#.

What I want to express is that I hope C# can simplify its syntax a bit. After all, ordinary programmers (such as me) don't care about its implementation process at all, just consider how much syntax sugar they have when writing.

...And so on.

But I found that these proposals were rarely accepted. I can understand the resistance of some people who have been using C# for a long time to these new grammars, and I can also understand that some grammars may have conflicts and problems in parser logic.

I hope I can convince you with a little words and change this situation. ha-ha.

**Comment by HaloFour** [comment]
@SF-Simon 

> I hope I can persuade everyone with a little words and change this situation.

I'm suggesting that you represent the change that you would like to see, open discussions around features or syntax that you would like to see C# implement.  Guidance like "be like Swift" isn't really actionable, but specific examples of syntax that you think would make development simpler could be considered.

**Comment by tilkinsc** [comment]
The proposal should clarify this circumstance,
```cs
if ((objName?.a = func()) == null)
{
	
}
```

In which null-aware assignments in an if statement condition will return:
* assignment lvalue is null (any variable?. is null), null
* assignment rvalue is null, null
* assignment lvalue/rvalue is not null, func()

Assignments still must follow the code of:
* if lvalue is null, rvalue is not evaluated, no assignment happens

Its important to keep the idea that null should be emplaced where `(objName?.a = func())` is:
```cs
if (null == null)
{

}
```

**Comment by Rekkonnect** [comment]
To me this looks implied from the semantics that the expression has when it's not an expression statement; meaning the expression's value is used. The same would apply to all contexts, including boolean variable assignments, if statements, while statements, do-while statements, for statements, etc.

The semantics are pretty clear for this case, even though it's probably best to never write code like that and split out the assignment's result into a variable and then compare it to null.

**Comment by jamesford42** [comment]
This is gold, can't tell you guys how many times I've written assign foo to bar if bar isn't null with ?. and had to remember it's not possible.

**Comment by Rekkonnect** [comment]
Are there any chances this is being included in C# 13? We've been waiting for this improvement for quite a long time and it's a massive QoL. Last time this was touched by the LDM was in 2022.

**Comment by 333fred** [comment]
Very unlikely at this point.
